### ISSUE TITLE
Spy objects with properties 

### ISSUE THREAD

**Issue by peterwdj** [jasmine/jasmine#1474]
I'm currently working on a tech test, and would like to be able to create a spy object that has properties, as a JS constructor function might have, in order to be able to properly isolate tests. I feel this functionality would be a useful addition to the [spyOnProperty functionality](https://stackoverflow.com/questions/33148496/how-can-i-spy-on-a-getter-property-using-jasmine/43793646#43793646) that was released relatively recently.

## Expected Behavior
When writing a test, it should be possible to create a spy object with properties that can be accessed using `constructorFunction.property`, as you might call `constructorFunction.method()` on a Jasmine spy object that had been created as follows (from the [Jasmine documentation](https://jasmine.github.io/2.0/introduction.html): 

```
beforeEach(function() {
    tape = jasmine.createSpyObj('tape', ['play', 'pause', 'stop', 'rewind']);

    tape.play();
    tape.pause();
    tape.rewind(0);
  });
```

## Current Behavior
As things stand currently, I have been unable to find another way to create a Jasmine spy object that holds properties as well as methods. This means it is not possible to isolate tests for functions that alter the properties of a different constructor (see below for example).

## Possible Solution
One possible implementation that suggests itself to me is to add a second, optional array of properties as key/value pairs when creating a Jasmine spy object. In practice, this would look something like this:

```
beforeEach(function() {
    tape = jasmine.createSpyObj('tape', ['play', 'pause', 'stop', 'rewind'], ['title': 'Guardians of the Galaxy Awesome Mix: Vol. 1', 'artist': 'Various artists', 'tracks': 12, 'playingTime': 42]);

    tape.play();
    tape.pause();
    tape.rewind(0);

    console.log(tape.title); // Guardians of the Galaxy Awesome Mix: Vol. 1 
  });
```

## Suite that reproduces the behavior (for bugs)
This is code to write a simple command-line banking application.

```
// account.js
'use strict';

var Deposit = require('./deposit.js').Deposit;
var Withdrawal = require('./withdrawal.js').Withdrawal;
const DEFAULT_BALANCE = 0;

var Account = function() {
  this._balance = DEFAULT_BALANCE;
  this._transactions = [];
};

Object.defineProperty(Account.prototype, 'balance', {
  get: function() { return this._balance },
  set: function(newBalance) { this._balance = newBalance },
  configurable: true
});

Account.prototype.deposit = function(depositAmount) {
  var deposit = new Deposit(this, depositAmount);
  deposit.deposit();
};

// More code
```
```
// deposit.js
'use strict';

var Deposit = function(account, depositAmount) {
  this.date = new Date();
  this.account = account;
  this.depositAmount = depositAmount;
};

Deposit.prototype.deposit = function() {
  var currentBalance = this.account.balance;
  var newBalance = currentBalance + this.depositAmount;
  this.account.balance = newBalance;
};

exports.Deposit = Deposit;
```
```
// DepositSpec.js
'use strict';

describe('Deposit', function() {
  var Account = require('../lib/account.js').Account;
  var Deposit = require('../lib/deposit.js').Deposit;
  var account;
  var deposit;

  beforeEach(function() {
    account = new Account();
    deposit = new Deposit(account, 50);
  })

  describe('Date', function() {
    it('Should be a date object', function() {
      expect(deposit.date).toEqual(jasmine.any(Date));
    });
  });

  describe('#deposit', function() {
    it('Should reduce an account\'s balance', function() {
      spyOnProperty(account, 'balance', 'get').and.returnValue(0);
      var balanceSetterSpy = spyOnProperty(account, 'balance', 'set').and.returnValue(50);
      account.deposit(50);
      expect(balanceSetterSpy).toHaveBeenCalled();
    });
  });
});
```

## Context
In the final test above, it is necessary to create a real account object in order to properly test the functionality of the deposit class, as it modifies the balance property of the account. However, in creating this real object, the test is no longer isolated. 

## Your Environment
* Version used: 2.8.0
* Ubuntu 16.04 run on a Chromebook Pixel via [crouton](https://github.com/dnschneid/)
* Node version: 4.2.6
* Link to your project: https://github.com/peterwdj/bankTechTest

**Comment by slackersoft** [comment]
This seems like something that could be interesting, but I'm not sure I fully see the utility. I'm not sure if these questions are because of a simplified example, or just I'm misunderstanding the full problem.

* Your usage of the property `balance` doesn't seem to truly need to be a defined property, probably just setting `balance = 0` and checking `expect(balance).toEqual(50)` should work fine.

* Your test is calling `account.deposit` so you would need to have a fully created account regardless of spied on properties.

Outside of tracking calls, the primary thing that a spy provides in your test is the ability to define what the return value is and check things on it. If you provide an object with a pre-defined value and then check the value at the end, this seems to provide the same benefit.

Hope this helps. Thanks for using Jasmine!

**Comment by peterwdj** [comment]
Hey Slackersoft, thanks for replying! Having spent more time with my code, I've come to the realisation that the approach I took to this problem was actually rather a poor one in terms of design choices, and I imagine this is part of the reason I struggled with this. 

In response to your two points:
- I don't think this would have worked, as balance was a property of account, which is what I was attempting to check (although in hindsight, that this was in the tests for my Deposit spec should have told me that this approach was an ill-thought-through one)
- Would it not have been possible to replicate that using a Jasmine spy, though, were this functionality to have existed?

I guess the main problem I see with using a simple (spy) object is how that value is accessed by the program. But no matter now, and from what I can see, this does not seem to be something that many other people have had issues with in the past, so no problems at all if you want to go ahead and close this issue :-) 

Thanks again for taking the time to respond!
